src/privileged/trackers/api.js

```diff
@@ -0,0 +1,269 @@
+          addExceptionButton.removeEventListener("command", this.onAddExceptionButtonCommand);
+        },
+        async trackerCallback(e) {
+          const tabId = tabTracker.getBrowserTabId(e.target);
+          let etld;
+          try {
+            etld = Services.eTLD.getBaseDomainFromHost(e.data.hostname);
+          } catch (error) {
+            return;
+          }
+          trackersEventEmitter.emitReloadWithTrackers(tabId, etld);
+        },
+        async pageErrorCallback(e) {
+          const tabId = tabTracker.getBrowserTabId(e.target);
+          trackersEventEmitter.emitErrorDetected(e.data, tabId);
+        },
+        async pageBeforeUnloadCallback(e) {
+          const tabId = tabTracker.getBrowserTabId(e.target);
+          let uri;
+          try {
+            uri = Services.io.newURI(e.data.telemetryData.completeLocation);
+            // Browser is never private, so type can always be "trackingprotection"
+            e.data.telemetryData.etld =
+              Services.eTLD.getBaseDomainFromHost(e.data.telemetryData.hostname);
+          } catch (error) {
+            return;
+          }
+          e.data.telemetryData.user_has_tracking_protection_exception =
+            Services.perms.testExactPermission(uri, "trackingprotection") === Services.perms.ALLOW_ACTION;
+          trackersEventEmitter.emitPageBeforeUnload(tabId, e.data.telemetryData);
+        },
+        async pageUnloadCallback(e) {
+          const tabId = tabTracker.getBrowserTabId(e.target);
+          trackersEventEmitter.emitPageUnload(tabId);
+        },
+        onReportBreakageButtonCommand() {
+          const win = BrowserWindowTracker.getTopWindow({
+            private: false,
+            allowPopups: false,
+          });
+          const tabId = tabTracker.getBrowserTabId(win.gBrowser.selectedBrowser);
+          trackersEventEmitter.emitReportBreakage(tabId);
+        },
+        async onAddExceptionButtonCommand() {
+          const win = BrowserWindowTracker.getTopWindow({
+            private: false,
+            allowPopups: false,
+          });
+          const tabId = tabTracker.getBrowserTabId(win.gBrowser.selectedBrowser);
+          trackersEventEmitter.emitAddException(tabId);
+        },
+        async setListeners(win) {
+          const mm = win.getGroupMessageManager("browsers");
+          // Web Progress Listener has detected a change.
+          mm.addMessageListener("reload-with-trackers", this.trackerCallback);
+          mm.addMessageListener("pageError", this.pageErrorCallback);
+          // We pass "true" as the third argument to signify that we want to listen
+          // to messages even when the framescript is unloading, to catch tabs closing.
+          mm.addMessageListener("beforeunload", this.pageBeforeUnloadCallback, true);
+          mm.addMessageListener("unload", this.pageUnloadCallback, true);
+
+          mm.loadFrameScript(context.extension.getURL("privileged/trackers/framescript.js"), true);
```

Comment:
Just use this.framescripturl here?

Reply:
So, there seems to be some weird race? condition where this.frameScriptURL is still undefined when that function is called. I don't know what weird webextension API sorcery is at play here, but I'd rather not touch it.

Good suggestion, though...